### PR TITLE
Fix Corefile migration logic

### PR DESCRIPTION
**Pull request by rajansandeep**

### CODE REVIEW

File: cmd/kubeadm/app/phases/addons/dns/dns.go [hunk 1]
```diff
@@ -280,6 +280,7 @@ func createCoreDNSAddon(deploymentBytes, serviceBytes, configBytes []byte, clien
 	canMigrateCorefile := true
 
 	if corefileMigrationRequired {
+		//If migration is required, try and migrate the Corefile
```
Comment (chrisohaver) [inline cmd/kubeadm/app/phases/addons/dns/dns.go]: We get in here for default corefiles that would require migration (e.g. when new defaults are added, or default options removed)... but we should instead be replacing those with the current default.

File: cmd/kubeadm/app/phases/addons/dns/dns.go [hunk 2]
```diff
@@ -289,10 +290,16 @@ func createCoreDNSAddon(deploymentBytes, serviceBytes, configBytes []byte, clien
 				return err
 			}
 		}
-	} else {
+	} else if isCoreDNSCorefileEmptyOrDefault(corefile) {
+		// If the Corefile is empty or default, the latest default Corefile will be applied
```
Comment (chrisohaver) [inline cmd/kubeadm/app/phases/addons/dns/dns.go]: I think the only time we get in here is when the default corefile is empty or already identical to the new default corefile (no migrations required).

File: cmd/kubeadm/app/phases/addons/dns/dns.go [hunk 3]
```diff
@@ -367,12 +374,17 @@ func createDNSService(dnsService *v1.Service, serviceBytes []byte, client client
 	return nil
 }
 
+// isCoreDNSCorefileEmptyOrDefault checks if the Corefile is empty or a unmodified Corefile
+func isCoreDNSCorefileEmptyOrDefault(corefile string) bool {
```
Comment (chrisohaver) [inline cmd/kubeadm/app/phases/addons/dns/dns.go]: I don't think this needs to be a separate function.

**Comment by chrisohaver** [comment]
Regarding replacing vs migrating a default corefile... it seems a minor point since essentially the end result is the same.  
I think replacing might be safer long term because of the possibility of an errant migration migrating default users "off" the default track.  I lean toward replacing over always migrating.
However, it would be simpler code wise to just always migrate in which case we never need to check too see if it's default or not. 

What do you think?

**Comment by rajansandeep** [comment]
> Regarding replacing vs migrating a default corefile... it seems a minor point since essentially the end result is the same.
> I think replacing might be safer long term because of the possibility of an errant migration migrating default users "off" the default track. I lean toward replacing over always migrating.
> However, it would be simpler code wise to just always migrate in which case we never need to check too see if it's default or not.
> 
> What do you think?

I think I agree that if the Corefile is default, we should just replace it rather than migrating it.
I think we can achieve that by not checking for `newdefaults`, since the default check includes any version of CoreDNS and should end up just replacing it when upgrading.

I'll add my next commit with debug statements and test results to see if it makes sense.

**Comment by rajansandeep** [comment]
The test results show as below:
```
=== RUN   TestCreateCoreDNSAddon
=== RUN   TestCreateCoreDNSAddon/Empty_Corefile
Updating the Corefile
=== RUN   TestCreateCoreDNSAddon/Default_Corefile
Updating the Corefile
=== RUN   TestCreateCoreDNSAddon/Remove_Deprecated_options
Migrating the Corefile
[addons] Migrating CoreDNS Corefile
=== RUN   TestCreateCoreDNSAddon/Update_proxy_plugin_to_forward_plugin
Migrating the Corefile
[addons] Migrating CoreDNS Corefile
=== RUN   TestCreateCoreDNSAddon/Modified_Corefile_with_no_migration_required
Retaining the Corefile
--- PASS: TestCreateCoreDNSAddon (0.01s)
    --- PASS: TestCreateCoreDNSAddon/Empty_Corefile (0.01s)
    --- PASS: TestCreateCoreDNSAddon/Default_Corefile (0.00s)
    --- PASS: TestCreateCoreDNSAddon/Remove_Deprecated_options (0.00s)
    --- PASS: TestCreateCoreDNSAddon/Update_proxy_plugin_to_forward_plugin (0.00s)
    --- PASS: TestCreateCoreDNSAddon/Modified_Corefile_with_no_migration_required (0.00s)
PASS
```

**Comment by chrisohaver** [comment]
I think default Corefiles are getting migrated, but the test doesn't care

Can you add a test case where the corefile is the default, but plugins not in the same order as the default (`migration.Default` should still return true in that case - it's not order sensitive).  If the Corefile is getting replaced, the outcome would be plugins ordered in the default order.  If it's getting migrated, it would retain the altered order of plugins from the original Corefile.  The test also needs to migrate to a version where the default Corefile has changed.

**Comment by chrisohaver** [comment]
> I think we can achieve that by not checking for newdefaults, since the default check includes any version of CoreDNS and should end up just replacing it when upgrading.

I think this would cause custom Corefiles not to get new defaults added to them, if there are no "removed" deprecations.

**Comment by rajansandeep** [comment]
> Can you add a test case where the corefile is the default, but plugins not in the same order as the default (migration.Default should still return true in that case - it's not order sensitive). If the Corefile is getting replaced, the outcome would be plugins ordered in the default order. If it's getting migrated, it would retain the altered order of plugins from the original Corefile. The test also needs to migrate to a version where the default Corefile has changed.

I added the test, it needed a change in execution order as you'd suggested earlier.
The shuffled default is replaced and the custom Corefile with "removed" deprecations is migrated with newdefault options

**Comment by chrisohaver** [comment]
> > I think we can achieve that by not checking for newdefaults, since the default check includes any version of CoreDNS and should end up just replacing it when upgrading.
> 
> I think this would cause custom Corefiles not to get new defaults added to them, if there are no "removed" deprecations.

I think this is still an issue.  For a non-default Corefile with no removed deprecations, `isCoreDNSConfigMapMigrationRequired` will return false.  It will then retain the Corefile, thus no new defaults would be added if there were any.

**Comment by chrisohaver** [comment]
> I think this is still an issue. For a non-default Corefile with no removed deprecations, isCoreDNSConfigMapMigrationRequired will return false. It will then retain the Corefile, thus no new defaults would be added if there were any.

A test for this would be to craft a non-default Corefile that has "removed" options pre-removed from it (e.g. `upstream`), but doesn't have the new defaults (e.g. `ready`, `max_concurrent`).  The expectation is that we would add the new defaults to the Corefile.

**Comment by rajansandeep** [comment]
> I think this is still an issue. For a non-default Corefile with no removed deprecations, `isCoreDNSConfigMapMigrationRequired` will return false. It will then retain the Corefile, thus no new defaults would be added if there were any.

You're right, I reverted the `newdefault` check for this and added a test case covering this.

**Comment by chrisohaver** [comment]
Looks good.  One nit: I think that `isCoreDNSCorefileEmptyOrDefault` is too simple to be a separate function that is only used once, and considering it could be a one line function, e.g. ...

```
func isCoreDNSCorefileEmptyOrDefault(corefile string) bool {
 	return corefile == "" || migration.Default("", corefile)
 }
```

**Comment by rajansandeep** [comment]
> Looks good. One nit: I think that `isCoreDNSCorefileEmptyOrDefault` is too simple to be a separate function that is only used once, and considering it could be a one line function, e.g. ...
> 
> ```
> func isCoreDNSCorefileEmptyOrDefault(corefile string) bool {
>  	return corefile == "" || migration.Default("", corefile)
>  }
> ```

cool! I removed it, if it looks good, I'll squash and send the PR upstream

**Comment by chrisohaver** [comment]
Looks good!

**Pull request merged**